moveit_core/planning_scene/src/planning_scene.cpp

```diff
@@ -411,58 +417,146 @@ void PlanningScene::checkCollision(const collision_detection::CollisionRequest&
   checkCollision(req, res, robot_state, getAllowedCollisionMatrix());
 }
 
-void PlanningScene::checkSelfCollision(const collision_detection::CollisionRequest& req,
-                                       collision_detection::CollisionResult& res)
+void PlanningScene::checkCollision(const collision_detection::CollisionRequest& req,
+                                   collision_detection::CollisionResult& res, moveit::core::RobotState& robot_state,
+                                   const collision_detection::AllowedCollisionMatrix& acm) const
 {
-  if (getCurrentState().dirtyCollisionBodyTransforms())
+  if (robot_state.dirtyCollisionBodyTransforms())
   {
-    checkSelfCollision(req, res, getCurrentStateNonConst());
-  }
-  else
-  {
-    checkSelfCollision(req, res, getCurrentState());
+    robot_state.updateCollisionBodyTransforms();
   }
+  checkCollision(req, res, static_cast<const moveit::core::RobotState&>(robot_state), acm);
 }
 
 void PlanningScene::checkCollision(const collision_detection::CollisionRequest& req,
                                    collision_detection::CollisionResult& res,
                                    const moveit::core::RobotState& robot_state,
-                                   const collision_detection::AllowedCollisionMatrix& acm) const
+                                   const collision_detection::AllowedCollisionMatrix& acm,
+                                   bool validate_transforms) const
 {
+  if (validate_transforms && robot_state.dirtyCollisionBodyTransforms())
+  {
+    RCLCPP_WARN(getLogger(), "Robot state has dirty collision body transforms. Please update the collision body "
+                             "transforms before checking collision.");
+  }
   // check collision with the world using the padded version
-  getCollisionEnv()->checkRobotCollision(req, res, robot_state, acm);
+  req.pad_environment_collisions ? getCollisionEnv()->checkRobotCollision(req, res, robot_state, acm) :
+                                   getCollisionEnvUnpadded()->checkRobotCollision(req, res, robot_state, acm);
+
+  // Return early if a collision was found or the maximum number of allowed contacts is exceeded
+  if (res.collision || (req.contacts && res.contacts.size() >= req.max_contacts))
```

Comment:
I don't fully understand this check.
If there's a collision, it returns right away.
If there's no collision, wouldn't `res.contacts.size()` be zero?
So if there's no env collision, but `req.contacts==true` and `req.max_contacts==0`, wouldn't this return without checking self-collisions?

Reply:
:thinking: Good point, would it be better to check self-collisions first because that should always be evaluated? I did not change that part so I am not sure if there is a good reason why it is the other way around.

Reply:
Looking again, I think the previous check is correct, we either have to fix this negated version, or use the previous